File: 448. Find All Numbers Disappeared in an Array.py
# Submitted by : Aryan Singh_RN12MAY2023
# Time Complexity : O(n)
# Space Complexity : Average : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

class Solution:
    '''
    - will use the hashset concept to process in place. number is compared to it's index to know if missing
    - when iterating, the number's index is marked with negative value
    - in second iteration checked if any number positive whose index then added to output list 
    '''
    def findDisappearedNumbers(self, nums: List[int]) -> List[int]:
        output = []
        # to change the value of numbers index to negative as marker
        for i in range(len(nums)):
            index = abs(nums[i]) - 1
            if nums[index] > 0:
                nums[index] *= -1
        # to find which number is not there
        for i in range(len(nums)):
            if nums[i] > 0:
                output.append(i + 1)
        
        return output

File: _448_FindAllNumbersDisappearedinanArray.java
// space complexity - o(1)
    //time - o(n)
    //Ran on leeetcode successfully : Yes
    // Problem faced  : No
    //Approach : -ve marker approach, use value of array element as index and mark element at that index as negative
class Solution {
    public List<Integer> findDisappearedNumbers(int[] nums) {
        List<Integer> l = new ArrayList<>();

        for(int i=0; i<nums.length; i++){
            int index  = Math.abs(nums[i]);
            if(nums[index-1] > 0){
                nums[index-1] *= -1;
            }
        }
        for(int i=0; i<nums.length; i++){
            if(nums[i] > 0)
                l.add(i+1);
        }
        return l;
    }
}
